### PR TITLE
Auto-correct command line tests

### LINKED ISSUE #6477: [Tests] Auto-correct commandline tests
It would be nice if the commandline tests (those that test input and output of solc and solc standard-json) could auto-update the expectations just like isoltest does.

### PR DESCRIPTION
**Pull request by sifmelcara**
### Description

Closes #6477 

When expectation mismatches, ask user if he/she wants to update the expectation file. Note that we don't consider different number of trailing newlines as "mismatch" because bash's command substitution removes them.

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### CODE REVIEW

File: test/cmdlineTests.sh [hunk 1]
```diff
@@ -133,37 +151,44 @@ function test_solc_behaviour()
...
     then
         printError "Incorrect exit code. Expected $exit_code_expected but got $exitCode."
-        rm -f $stdout_path $stderr_path
         exit 1
     fi
 
     if [[ "$(cat $stdout_path)" != "${stdout_expected}" ]]
     then
         printError "Incorrect output on stdout received. Expected:"
-        echo -e "${stdout_expected}"
```
Comment (sifmelcara) [inline test/cmdlineTests.sh:143]: I'm not sure if this `-e` have any purpose so I removed it.

File: test/cmdlineTests.sh [hunk 2]
```diff
@@ -133,37 +151,44 @@ function test_solc_behaviour()
...
+            ask_expectation_update "$(cat $stdout_path)" "$stdout_expectation_file"
+        else
+            exit 1
+        fi
     fi
 
     if [[ "$(cat $stderr_path)" != "${stderr_expected}" ]]
     then
         printError "Incorrect output on stderr received. Expected:"
-        echo -e "${stderr_expected}"
```
Comment (sifmelcara) [inline test/cmdlineTests.sh:156]: Also here

**Comment by sifmelcara** [comment]
Force pushed to add user input timeout in case of expectation mismatch on CI

File: test/cmdlineTests.sh [hunk 3]
```diff
@@ -93,6 +93,27 @@ function compileFull()
...
 }
 
+function ask_expectation_update()
+{
+    local newExpectation="${1}"
+    local expectationFile="${2}"
+    while true;
+    do
+        set +e
+        read -t10 -p "(u)pdate expectation/(q)uit? "
```
Comment (erak) [inline test/cmdlineTests.sh:103]: There's an additional whitespace at the end.

Comment (sifmelcara) [reply test/cmdlineTests.sh:103]: It is intentional, I copied the string from isoltest https://github.com/ethereum/solidity/blob/ef3a18999c840c523cce73f6074e816334b5c753/test/tools/isoltest.cpp#L179

File: test/cmdlineTests.sh [hunk 4]
```diff
@@ -210,14 +243,14 @@ printTask "Testing unknown options..."
 
 
 printTask "Testing passing files that are not found..."
-test_solc_behaviour "file_not_found.sol" "" "" "" 1 "\"file_not_found.sol\" is not found."
+test_solc_behaviour "file_not_found.sol" "" "" "" 1 "\"file_not_found.sol\" is not found." "" ""
```
Comment (erak) [inline test/cmdlineTests.sh:246]: Why are these filepaths empty?

Comment (sifmelcara) [reply test/cmdlineTests.sh:246]: Because they don't have their expected outputs saved as files, we cannot auto-update it.(those 2 empty arguments are `stdout_expectation_file` and `stderr_expectation_file`)

Comment (erak) [reply test/cmdlineTests.sh:246]: Ah, ok. I see.

File: test/cmdlineTests.sh [hunk 5]
```diff
@@ -93,6 +93,27 @@ function compileFull()
...
+{
+    local newExpectation="${1}"
+    local expectationFile="${2}"
+    while true;
+    do
+        set +e
+        read -t10 -p "(u)pdate expectation/(q)uit? "
+        if [ $? -gt 128 ];
+        then
+            echo -e "\nuser input timed out"
```
Comment (erak) [inline test/cmdlineTests.sh]:
```suggestion
            echo -e "\nUser input timed out."
```

File: test/cmdlineTests.sh (hunk 1 above)
Comment (erak) [reply test/cmdlineTests.sh:143]: I think it's fine to remove it. I also don't see a need for that at the moment.

File: test/cmdlineTests.sh [hunk 6]
```diff
@@ -133,37 +159,44 @@ function test_solc_behaviour()
...
     if [[ "$(cat $stdout_path)" != "${stdout_expected}" ]]
     then
         printError "Incorrect output on stdout received. Expected:"
-        echo -e "${stdout_expected}"
+        echo "${stdout_expected}"
 
         printError "But got:"
         cat $stdout_path
+
         printError "When running $SOLC ${filename} ${solc_args} <$solc_stdin"
```
Comment (erak) [inline test/cmdlineTests.sh]: Not really part of this PR, but the output looks weird, if `$solc_stdin` is empty. Could you also add a condition for printing the additional `<`?

**Comment by erak** [comment]
Looks good so far. @sifmelcara Thanks for this PR!

File: test/cmdlineTests.sh (hunk 5 above)
Comment (sifmelcara) [reply test/cmdlineTests.sh]: Done

File: test/cmdlineTests.sh (hunk 6 above)
Comment (sifmelcara) [reply test/cmdlineTests.sh]: Ok I changed how it print out the command

File: test/cmdlineTests.sh [hunk 7]
```diff
@@ -103,17 +124,22 @@ function test_solc_behaviour()
...
+    local stderr_expectation_file="${8}" # the file to write to when user chooses to update stderr expectation
     local stdout_path=`mktemp`
     local stderr_path=`mktemp`
+
+    trap "rm -f $stdout_path $stderr_path" EXIT
+
     if [[ "$exit_code_expected" = "" ]]; then exit_code_expected="0"; fi
 
+    local solc_command="$SOLC ${filename} ${solc_args}"
+    if [[ -n "$solc_stdin"  ]]; then solc_command+=" <$solc_stdin"  ; fi
```
Comment (erak) [inline test/cmdlineTests.sh:137]: Is it enough to quote `solc_stdin` in order to not make this script vulnerable in the context of `eval` below?

File: test/cmdlineTests.sh [hunk 8]
```diff
@@ -133,37 +159,44 @@ function test_solc_behaviour()
     if [[ $exitCode -ne "$exit_code_expected" ]]
     then
         printError "Incorrect exit code. Expected $exit_code_expected but got $exitCode."
-        rm -f $stdout_path $stderr_path
         exit 1
     fi
 
     if [[ "$(cat $stdout_path)" != "${stdout_expected}" ]]
```
Comment (erak) [inline test/cmdlineTests.sh:165]: Could this block and the one starting in line 183 generalized into a function?

File: test/cmdlineTests.sh (hunk 7 above)
Comment (sifmelcara) [reply test/cmdlineTests.sh:137]: Did you mean that someone will create some directories with strange directory name to execute malicious shell command on solidity developer's computer?

File: test/cmdlineTests.sh (hunk 8 above)
Comment (sifmelcara) [reply test/cmdlineTests.sh:165]: Yes, but maybe we can do it later when we also want auto-correction for yul command line test, etc?

File: test/cmdlineTests.sh (hunk 7 above)
Comment (sifmelcara) [reply test/cmdlineTests.sh:137]: (Asked the above question because I couldn't catch what you said)

Comment (erak) [reply test/cmdlineTests.sh:137]: Yes, I wasn't thinking about weird directory names in particular, but rather about a weird standard input. But that should not be problem, since it's only passed internally. As far as I can see, the filenames etc. are properly quoted. So it should be fine.

File: test/cmdlineTests.sh (hunk 8 above)
Comment (erak) [reply test/cmdlineTests.sh:165]: Sounds good!

**Review by erak — APPROVED** [review]

**Review by ekpyron — APPROVED** [review]

File: test/cmdlineTests.sh [hunk 9]
```diff
@@ -133,37 +159,44 @@ function test_solc_behaviour()
...
     then
         printError "Incorrect output on stdout received. Expected:"
-        echo -e "${stdout_expected}"
+        echo "${stdout_expected}"
```
Comment (chriseth) [inline test/cmdlineTests.sh:168]: I think it does format some json output a bit more nicely, if I remember correctly.
